xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp

```diff
@@ -0,0 +1,543 @@
+/*
+ *  Copyright (C) 2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DVDVideoCodecStarfish.h"
+
+#include "CompileInfo.h"
+#include "DVDCodecs/DVDFactoryCodec.h"
+#include "ServiceBroker.h"
+#include "cores/VideoPlayer/Buffers/VideoBuffer.h"
+#include "cores/VideoPlayer/Interface/DemuxCrypto.h"
+#include "cores/VideoPlayer/Interface/TimingConstants.h"
+#include "cores/VideoPlayer/VideoRenderers/RenderManager.h"
+#include "media/decoderfilter/DecoderFilterManager.h"
+#include "messaging/ApplicationMessenger.h"
+#include "settings/Settings.h"
+#include "settings/SettingsComponent.h"
+#include "utils/BitstreamConverter.h"
+#include "utils/CPUInfo.h"
+#include "utils/TimeUtils.h"
+#include "utils/log.h"
+#include "windowing/wayland/WinSystemWaylandWebOS.h"
+
+#include <memory>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+using namespace KODI::MESSAGING;
+
+enum STARFISH_STATES
+{
+  STARFISH_STATE_RESET,
+  STARFISH_STATE_FLUSHED,
+  STARFISH_STATE_RUNNING,
+};
+
+/*****************************************************************************/
+/*****************************************************************************/
+CDVDVideoCodecStarfish::CDVDVideoCodecStarfish(CProcessInfo& processInfo)
+  : CDVDVideoCodec(processInfo),
+    m_starfishMediaAPI(std::make_unique<StarfishMediaAPIs>()),
+    m_state(STARFISH_STATE_FLUSHED)
+{
+}
+
+CDVDVideoCodecStarfish::~CDVDVideoCodecStarfish()
+{
+  Dispose();
+}
+
+std::unique_ptr<CDVDVideoCodec> CDVDVideoCodecStarfish::Create(CProcessInfo& processInfo)
+{
+  return std::make_unique<CDVDVideoCodecStarfish>(processInfo);
+}
+
+bool CDVDVideoCodecStarfish::Register()
+{
+  CDVDFactoryCodec::RegisterHWVideoCodec("starfish_dec", &CDVDVideoCodecStarfish::Create);
+  return true;
+}
+
+std::atomic<bool> CDVDVideoCodecStarfish::m_InstanceGuard(false);
+
+bool CDVDVideoCodecStarfish::Open(CDVDStreamInfo& hints, CDVDCodecOptions& options)
+{
+  m_opened = false;
+  std::string payload;
+  auto payloadArgs = nlohmann::json();
+  auto payloadArg = nlohmann::json();
+
+  // allow only 1 instance here
+  if (m_InstanceGuard.exchange(true))
+  {
+    CLog::Log(LOGERROR, "CDVDVideoCodecStarfish::Open - InstanceGuard locked");
+    return false;
+  }
+
+  if (!hints.width || !hints.height)
+  {
+    CLog::Log(LOGERROR, "CDVDVideoCodecStarfish::Open - {}", "null size, cannot handle");
+    m_InstanceGuard.exchange(false);
```

Comment:
this is error-prone, please arrange this so that you cannot forget resetting ms_instanceGuard on error paths

Reply:
Hmm, I have to think about this. Is that why there is a GOTO in the Android implementation?

Reply:
ahhh, the Android implementation! I was really wondering where the atomic came from :D I discussed a bit with @lrusak and we're not sure if it's actually necessary, since we typically don't try to open multiple hwdecs anyway.
Yes, I'm sure this is was one of the reasons that there is `goto` in Android. We don't accept new code that uses goto, though ^^;

Reply:
Take care that on StreamChange Videoplayer would like to open the other stream quasi parallel to the current stream. Therefore the instance had to be guarded iirc.

Reply:
How is that supposed to work then? The second open would just fail if I'm reading it correctly.

Reply:
Yes - as you cannot assume to have multi decoder instances on Android. It fails - so player tears all down (including the running instance) and opens a new one. This is at least what I think to remember.

Reply:
I think there's TVs with only one decoder available so guarding only one instance seems safer

Reply:
@fritsch That seems a very roundabout way for us to do it, but certainly we don't want to reinvent VideoPlayer in this PR :)

@sundermann so wouldn't the second open just "naturally" fail in that case?

Reply:
Not sure, any idea how we can deliberately open a lot of decoding instances? I have 8 vdec devices in `/dev/lg/vdec*` so not sure if I would be able to trigger the limitation on my TV

Reply:
Also when would a stream change happen? Adaptive streaming?

Reply:
Channel Switch on PVR.
LiveTV move -> commercials. Here audio format, aspect ratio and stuff changes. While Android has a ReconfigureDecoder API other OSs have not and you need to reopen.

One does not want to see a loading spinner when this happens.

Reply:
I don't have a PVR server to test. But I think that right now parallel streams wouldn't work because we'd be using the same webos foreign exported window at the same time for a short period of time.

Hmm, I don't think we can reconfigure the decoder. Does that mean we should return false in `Reconfigure`?

Reply:
Guarding happens now in `Open` and the actual code has been moved into a new function `OpenInternal`.